common/hyperscaler/rules/rule.go

```diff
@@ -0,0 +1,224 @@
+		return plan
+	} else if plan == "trial" {
+		return "aws"
+	} else if plan == "free" {
+		return "aws/azure"
+	} else if plan == "sap-converged-cloud" {
+		return "openstack"
+	} else if plan == "preview" {
+		return "aws"
+	} else {
+		return ""
+	}
+}
+
+func (r *Rule) Matched(attributes *MatchableAttributes) bool {
+
+	if r.Plan != attributes.Plan {
+		return false
+	}
+
+	matched := true
+	for _, attr := range InputAttributes {
+		value := attr.Getter(r)
+		matchableValue := attr.MatchableGetter(attributes)
+		matched = matched && (value == matchableValue || value == ASTERISK || value == "")
+	}
+
+	return matched
+}
+
+func (r *Rule) SetAttributeValue(attribute, value string) (*Rule, error) {
+	for _, attr := range AllAttributes {
+		if attr.Name == attribute {
+			return attr.Setter(r, value)
+		}
+	}
+
+	return nil, fmt.Errorf("unknown attribute %s", attribute)
+}
+
+func (r *Rule) NumberOfInputAttributes() int {
+	count := 0
+
+	for _, attr := range InputAttributes {
+		value := attr.Getter(r)
+		if value != "" {
+			count++
+		}
+	}
+
+	return count
+}
+
+func (r *Rule) String() string {
+	ruleStr := r.StringNoLabels()
+
+	labels := r.CalculateLabels()
+	labelsStr := "# "
+	labelsToSort := make([]string, 0, len(labels))
+	for key, value := range labels {
+		labelsToSort = append(labelsToSort, fmt.Sprintf("%s: %s", key, string(value)))
+	}
+
+	sort.Strings(labelsToSort)
+
+	for _, key := range labelsToSort {
+		labelsStr += fmt.Sprintf("%s, ", key)
+	}
+
+	labelsStr = labelsStr[:len(labelsStr)-2]
+
+	return fmt.Sprintf("%-50s %-50s", ruleStr, labelsStr)
+}
+
+func (r *Rule) StringNoLabels() string {
+	ruleStr := fmt.Sprintf("%s", r.Plan)
+
+	if r.ContainsInputAttributes {
+
+		ruleStr += fmt.Sprintf("(")
+
+		for _, attr := range InputAttributes {
+			attrStr := attr.String(r)
+			ruleStr += fmt.Sprintf("%s", attrStr)
+		}
+
+		ruleStr = ruleStr[:len(ruleStr)-2]
+		ruleStr += fmt.Sprintf(")")
+	}
+
+	if r.ContainsOutputAttributes {
+		ruleStr += fmt.Sprintf("-> ")
+
+		for _, attr := range OutputAttributes {
+			attrStr := attr.String(r)
+			ruleStr += fmt.Sprintf("%s", attrStr)
+		}
+
+		ruleStr = ruleStr[:len(ruleStr)-2]
+	}
+
+	return ruleStr
+}
+
+func (r *Rule) IsResolved() bool {
+	resolved := true
+
+	for _, attr := range InputAttributes {
+		value := attr.Getter(r)
+		resolved = resolved && value != ASTERISK
+	}
+
+	return resolved
+}
+
+func (r *Rule) Combine(rule Rule) *Rule {
+	newRule := NewRule()
+	newRule.SetPlan(r.Plan)
+
+	for _, attr := range InputAttributes {
+		value := attr.Getter(r)
+		if value != "" && value != ASTERISK {
+			attr.Setter(newRule, value)
+			newRule.ContainsInputAttributes = true
+		} else {
+			valueR := attr.Getter(&rule)
+			attr.Setter(newRule, valueR)
+			newRule.ContainsInputAttributes = true
+		}
+	}
+
+	newRule.hyperscalerNameMappingFunction = r.hyperscalerNameMappingFunction
+
+	return newRule
+}
+
+func (r *Rule) SignatureWithValues() string {
```

Comment:
If we compute SignatureWithValues why we name the result `key`?

Reply:
https://github.com/ralikio/kyma-environment-broker/commit/fb4019460ee776ffd8fd93148493d5a37797ffe7